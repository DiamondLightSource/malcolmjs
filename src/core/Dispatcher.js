/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
