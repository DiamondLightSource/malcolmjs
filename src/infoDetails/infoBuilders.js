/* eslint no-underscore-dangle: 0 */
import { AlarmStates } from '../malcolmWidgets/attributeDetails/attributeAlarm/attributeAlarm.component';

// eslint-disable-next-line import/prefer-default-export
export const buildAttributeInfo = props => {
  let value;
  const info = {};
  const { attribute } = props;
  if (attribute && attribute.raw && attribute.raw.meta) {
    if (props.subElement === undefined) {
      info.path = {
        label: 'Attribute path',
        value: `${props.attribute.calculated.path[0]}, ${
          props.attribute.calculated.path[1]
        }`,
        inline: true,
      };
      info.meta = {
        label: 'Meta Data',
        malcolmType: {
          valuePath: 'raw.meta.typeid',
          label: 'Malcolm Type',
          inline: true,
        },
        description: {
          valuePath: 'raw.meta.description',
          label: 'Description',
          inline: true,
        },
        writeable: {
          valuePath: 'raw.meta.writeable',
          label: 'Writeable?',
          inline: true,
          tag: 'widget:led',
        },
      };
      info.malcolmAlarm = {
        label: 'Alarm',
      };
      Object.keys(attribute.raw.alarm).forEach(key => {
        info.malcolmAlarm[key] = {
          label: key,
          inline: true,
          valuePath: `raw.alarm.${key}`,
        };
        if (key === 'severity') {
          info.malcolmAlarm[key].alarmStatePath = 'calculated.alarms.rawAlarm';
        }
      });

      info.malcolmAlarm.message = {
        label: 'message',
        inline: true,
        valuePath: 'raw.alarm.message',
      };

      info.timeStamp = {
        label: 'Time Stamp',
        time: {
          label: 'time',
          inline: true,
          valuePath: 'calculated.timeStamp',
        },
      };
      Object.keys(attribute.raw.timeStamp).forEach(key => {
        info.timeStamp[key] = {
          label: key,
          inline: true,
          valuePath: `raw.timeStamp.${key}`,
        };
      });

      info.errorState = {
        label: 'Error State',
        valuePath: 'calculated.errorMessage',
        inline: true,
        alarmStatePath: 'calculated.alarms.errorState',
      };
      if (
        attribute.raw.meta.tags.some(a =>
          ['widget:table', 'widget:textinput'].includes(a)
        )
      ) {
        info.localState = {
          label: 'Local State',
          value: 'Discard',
          disabledPath: 'NOT.calculated.dirty',
          inline: true,
          tag: 'info:button',
          alarmStatePath: 'calculated.alarms.dirty',
        };
        if (props.revertHandler) {
          info.localState.functions = {
            clickHandler: () => {
              props.revertHandler(props.attribute.calculated.path);
            },
          };
        }
      }
      // eslint-disable-next-line prefer-destructuring
      value = attribute.raw.value;
    } else if (attribute.localState) {
      const row = parseInt(props.subElement[1], 10);
      const rowFlags = attribute.localState.flags.rows[row];
      const isNewRow =
        row >= attribute.raw.value[attribute.localState.labels[0]].length;
      info.localState = {
        label: 'Row local state',
        value: 'Discard',
        disabled: !(rowFlags._dirty || rowFlags._isChanged) || isNewRow,
        inline: true,
        tag: 'info:button',
        alarmState:
          rowFlags._dirty || rowFlags._isChanged ? AlarmStates.DIRTY : null,
      };
      const dataRow = {};
      attribute.localState.labels.forEach(label => {
        dataRow[label] =
          row < attribute.raw.value[label].length
            ? attribute.raw.value[label][row]
            : 'undefined';
      });
      info.rowValue = {
        label: 'Row remote state',
        ...dataRow,
      };
      info.addRowAbove = {
        label: 'Insert row above',
        value: 'Add',
        inline: true,
        tag: 'info:button',
      };
      info.addRowBelow = {
        label: 'Insert row below',
        value: 'Add',
        inline: true,
        tag: 'info:button',
      };
      info.deleteRow = {
        label: 'Delete row',
        value: 'Delete',
        inline: true,
        tag: 'info:button',
      };
      if (props.addRow) {
        info.addRowAbove.functions = {
          clickHandler: () => {
            props.addRow(props.attribute.calculated.path, row);
            props.changeInfoHandler(
              props.attribute.calculated.path,
              `row.${row + 1}`
            );
          },
        };
        info.addRowBelow.functions = {
          clickHandler: () => {
            props.addRow(props.attribute.calculated.path, row, 'below');
          },
        };
        info.deleteRow.functions = {
          clickHandler: () => {
            if (row >= props.attribute.localState.value.length - 1) {
              if (row !== 0) {
                props.changeInfoHandler(
                  props.attribute.calculated.path,
                  `row.${row - 1}`
                );
              } else {
                props.closeInfoHandler(props.attribute.calculated.path);
              }
            }
            props.addRow(props.attribute.calculated.path, row, 'delete');
          },
        };
      }
      if (props.rowRevertHandler) {
        info.localState.functions = {
          clickHandler: () => {
            props.rowRevertHandler(
              props.attribute.calculated.path,
              dataRow,
              row
            );
          },
        };
      }
      info.subElement = props.subElement;
    }
  } else if (
    attribute &&
    attribute.calculated &&
    attribute.calculated.isMethod
  ) {
    if (props.subElement === undefined) {
      info.path = {
        label: 'Attribute path',
        value: `${attribute.calculated.path[0]}, ${
          attribute.calculated.path[1]
        }`,
        inline: true,
      };
      info.meta = {
        label: 'Meta Data',
        malcolmType: {
          value: attribute.raw.typeid,
          label: 'Malcolm Type',
          inline: true,
        },
        description: {
          value: attribute.raw.description,
          label: 'Description',
          inline: true,
        },
        writeable: {
          value: attribute.raw.writeable,
          label: 'Writeable?',
          inline: true,
          tag: 'widget:led',
        },
      };
      info.takes = { label: 'Input parameter types' };
      Object.keys(attribute.raw.takes.elements).forEach(input => {
        info.takes[input] = {
          label: attribute.raw.takes.elements[input].label,
          value: attribute.raw.takes.elements[input].typeid,
          infoPath: {
            root: attribute.calculated.path,
            subElement: `takes.${input}`,
          },
        };
      });

      info.returns = { label: 'Output parameter types' };
      Object.keys(attribute.raw.returns.elements).forEach(input => {
        info.returns[input] = {
          label: attribute.raw.returns.elements[input].label,
          value: attribute.raw.returns.elements[input].typeid,
          infoPath: {
            root: attribute.calculated.path,
            subElement: `returns.${input}`,
          },
        };
      });
      info.errorState = {
        label: 'Error State',
        inline: true,
        value: attribute.calculated.errorState
          ? attribute.calculated.errorMessage
          : 'n/a',
        alarmState: attribute.calculated.errorState
          ? AlarmStates.MAJOR_ALARM
          : null,
      };
    } else {
      const rawInfo =
        attribute.raw[props.subElement[0]].elements[props.subElement[1]];
      Object.keys(rawInfo).forEach(key => {
        info[key] = rawInfo[key];
      });
    }
  }
  return { info, value };
};

export const linkInfo = props => {
  const portAttribute = props.attribute;

  if (!portAttribute || !portAttribute.raw.value) {
    return { ...props, info: {}, value: {} };
  }

  const blockMri = props.attribute.calculated.path[0];
  const blockName = props.linkBlockName;
  const portName = props.attribute.calculated.path[1];

  const portNullValue = portAttribute.raw.meta.tags
    .find(t => t.indexOf('inport:') > -1)
    .split(':')
    .slice(-1)[0];

  const info = {
    sourcePort: {
      label: 'Source',
      valuePath: 'raw.value',
      inline: true,
      tag: portAttribute.raw.meta.choices ? 'widget:combo' : 'widget:textinput',
      choices: portAttribute.raw.meta.choices,
      functions: {
        eventHandler: (nullPath, value) =>
          props.eventHandler([blockMri, portName, 'value'], value),
      },
    },
    destinationPort: {
      label: 'Destination',
      value: `${blockName}.${portName}`,
      inline: true,
    },
    deleteLink: {
      label: '',
      value: 'Delete',
      inline: true,
      tag: 'info:button',
      showLabel: false,
      functions: {
        clickHandler: () => {
          props.eventHandler([blockMri, portName, 'value'], portNullValue);
        },
      },
    },
  };

  return {
    info,
    value: {},
  };
};
