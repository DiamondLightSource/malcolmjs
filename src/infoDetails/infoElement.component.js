import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import AttributeAlarm, {
  AlarmStates,
} from '../malcolmWidgets/attributeDetails/attributeAlarm/attributeAlarm.component';
import { selectorFunction } from '../malcolmWidgets/attributeDetails/attributeSelector/attributeSelector.component';

import blockUtils from '../malcolm/blockUtils';

const styles = theme => ({
  div: {
    backgroundColor: theme.palette.background.paper,
    display: 'flex',
    paddingLeft: 5,
    paddingRight: 5,
    alignItems: 'center',
  },
  textName: {
    flexGrow: 1,
    textAlign: 'Left',
    marginLeft: 4,
    marginRight: 4,
  },
  controlContainer: {
    width: 150,
    padding: 2,
  },
  button: {
    width: '22px',
    height: '22px',
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
});

const InfoElement = props => (
  <div className={props.classes.div}>
    {props.alarm === AlarmStates.NO_ALARM ? (
      <IconButton className={props.classes.button} disableRipple>
        <AttributeAlarm alarmSeverity={props.alarm} />
      </IconButton>
    ) : (
      <AttributeAlarm alarmSeverity={props.alarm} />
    )}
    <Typography className={props.classes.textName}>
      {props.showLabel ? `${props.label}:` : ''}
    </Typography>
    <div className={props.classes.controlContainer}>
      {selectorFunction(
        props.tag,
        props.path,
        props.value,
        props.handlers.eventHandler,
        { isDisabled: props.disabled },
        props.handlers.setFlag,
        props.theme.palette.primary.light,
        { choices: props.choices },
        false,
        false,
        props.handlers.clickHandler
      )}
    </div>
  </div>
);

InfoElement.propTypes = {
  alarm: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
  showLabel: PropTypes.bool,
  tag: PropTypes.string.isRequired,
  path: PropTypes.arrayOf(PropTypes.string),
  value: PropTypes.oneOf([PropTypes.string, PropTypes.number, PropTypes.bool])
    .isRequired,
  choices: PropTypes.arrayOf(PropTypes.string),
  classes: PropTypes.shape({
    div: PropTypes.string,
    textName: PropTypes.string,
    controlContainer: PropTypes.string,
    button: PropTypes.string,
  }).isRequired,
  theme: PropTypes.shape({
    palette: PropTypes.shape({
      primary: PropTypes.shape({
        light: PropTypes.string,
      }),
      secondary: PropTypes.shape({
        main: PropTypes.string,
      }),
    }),
  }).isRequired,
  handlers: PropTypes.shape({
    eventHandler: PropTypes.func,
    setFlag: PropTypes.func,
    clickHandler: PropTypes.func,
  }),
  disabled: PropTypes.bool.isRequired,
};

InfoElement.defaultProps = {
  choices: [],
  handlers: {},
  path: [''],
  showLabel: true,
};

const mapStateToProps = (state, ownProps) => {
  let value;
  let alarmState;
  let disabled;
  const attribute = blockUtils.findAttribute(
    state.malcolm.blocks,
    ownProps.blockName,
    ownProps.attributeName
  );
  if (attribute) {
    if (ownProps.valuePath) {
      const path = ownProps.valuePath.split('.');
      value = attribute;
      path.forEach(pathElement => {
        value = value[pathElement];
      });
    }
    if (ownProps.alarmPath) {
      const path = ownProps.alarmPath.split('.');
      alarmState = attribute;
      path.forEach(pathElement => {
        alarmState = alarmState[pathElement];
      });
    }
    if (ownProps.disabledFlagPath) {
      const fullPath = ownProps.disabledFlagPath.split('.');
      const path = fullPath[0] === 'NOT' ? fullPath.slice(1) : fullPath;
      disabled = attribute;
      path.forEach(pathElement => {
        disabled = disabled[pathElement];
      });
      disabled = fullPath[0] === 'NOT' ? !disabled : !!disabled;
    }
  }
  return {
    alarm: alarmState !== undefined ? alarmState : ownProps.alarm,
    value: value !== undefined ? value : ownProps.value,
    disabled: disabled !== undefined ? disabled : ownProps.disabled,
  };
};

export default connect(mapStateToProps)(
  withStyles(styles, { withTheme: true })(InfoElement)
);
