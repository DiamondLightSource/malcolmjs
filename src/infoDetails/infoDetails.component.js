/* eslint react/no-array-index-key: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import GroupExpander from '../malcolmWidgets/groupExpander/groupExpander.component';
import InfoElement from './infoElement.component';
import { AlarmStates } from '../malcolmWidgets/attributeDetails/attributeAlarm/attributeAlarm.component';
import NavTypes from '../malcolm/NavTypes';
import {
  malcolmPutAction,
  malcolmSetFlag,
  malcolmRevertAction,
  malcolmUpdateTable,
  malcolmSelectLink,
  malcolmGetAction,
  clearError,
} from '../malcolm/malcolmActionCreators';
import { malcolmClearMethodInput } from '../malcolm/actions/method.actions';
import { buildAttributeInfo, linkInfo } from './infoBuilders';
import blockUtils from '../malcolm/blockUtils';
import navigationActions from '../malcolm/actions/navigation.actions';
import { idSeparator } from '../layout/layout.component';

const getTag = element => {
  if (element && element.tag) {
    return element.tag;
  }
  return 'info:multiline';
};

const infoAlarmState = value => {
  if (value instanceof Object) {
    if (value.alarmState !== undefined) {
      return value.alarmState;
    } else if (
      Object.keys(value).some(
        a =>
          ![
            'value',
            'valuePath',
            'alarmState',
            'alarmStatePath',
            'disabled',
            'disabledPath',
            'label',
            'showLabel',
            'tag',
            'inline',
            'functions',
            'choices',
          ].includes(a)
      )
    ) {
      return AlarmStates.NO_ALARM;
    }
  }
  return null;
};

const getValue = value => {
  if (!(value instanceof Object)) {
    return value;
  } else if (Object.prototype.hasOwnProperty.call(value, 'value')) {
    return value.value;
  } else if (value.valuePath !== undefined) {
    return undefined;
  }
  return Object.values(value)[0];
};

export class InfoDetails extends React.Component {
  static getDerivedStateFromProps(props, state) {
    if (
      props.blockName !== state.blockName ||
      props.attributeName !== state.attributeName ||
      (props.subElement && props.subElement !== state.subElement) ||
      (props.subElement && props.subElement[0] === 'row') ||
      (props.linkBlockName && props.linkBlockName !== state.subElement) ||
      props.isMethodInfo ||
      props.isLinkInfo ||
      Object.keys(state.info).length === 0 ||
      props.isLayoutLocked !== state.isLayoutLocked
    ) {
      let newState;
      let subElement;
      if (props.isLinkInfo) {
        subElement = props.linkBlockName;
        newState = linkInfo(props);
      } else {
        newState = buildAttributeInfo(props);
        // eslint-disable-next-line prefer-destructuring
        subElement = props.subElement;
      }
      return {
        ...newState,
        blockName: props.blockName,
        attributeName: props.attributeName,
        subElement,
        isLinkInfo: !!props.isLinkInfo,
        isLayoutLocked: props.isLayoutLocked,
      };
    }
    return state;
  }

  constructor(props) {
    super(props);
    this.state = {
      info: {},
    };
    if (props.info) {
      this.state.info = { ...props.info };
    }
  }

  shouldComponentUpdate(nextProps) {
    return (
      nextProps.blockName !== this.state.blockName ||
      nextProps.attributeName !== this.state.attributeName ||
      (!!nextProps.subElement &&
        (nextProps.subElement !== this.state.subElement ||
          nextProps.subElement[0] === 'row' ||
          nextProps.subElement[0] === 'col')) ||
      (!!nextProps.linkBlockName &&
        nextProps.linkBlockName !== this.state.subElement) ||
      nextProps.isMethodInfo ||
      nextProps.isLinkInfo ||
      Object.keys(this.state.info).length === 0 ||
      nextProps.isLayoutLocked !== this.props.isLayoutLocked
    );
  }

  render() {
    const infoElements = Object.keys(this.state.info).filter(
      a => this.state.info[a].inline || !(this.state.info[a] instanceof Object)
    );
    const infoGroups = Object.keys(this.state.info).filter(
      a => this.state.info[a] instanceof Object && !this.state.info[a].inline
    );
    return (
      <div>
        {infoElements.map(a => (
          <InfoElement
            key={a}
            label={this.state.info[a].label ? this.state.info[a].label : a}
            value={getValue(this.state.info[a])}
            valuePath={
              this.state.info[a].valuePath
                ? this.state.info[a].valuePath
                : undefined
            }
            alarm={infoAlarmState(this.state.info[a])}
            alarmPath={
              this.state.info[a].alarmStatePath
                ? this.state.info[a].alarmStatePath
                : undefined
            }
            tag={getTag(this.state.info[a])}
            handlers={this.state.info[a].functions}
            choices={this.state.info[a].choices}
            showLabel={this.state.info[a].showLabel}
            blockName={this.props.blockName}
            attributeName={this.props.attributeName}
            disabled={this.state.info[a].disabled}
            disabledFlagPath={this.state.info[a].disabledPath}
          />
        ))}
        {infoGroups.map((group, groupIndex) => (
          <GroupExpander
            key={groupIndex}
            groupName={
              this.state.info[group].label
                ? this.state.info[group].label
                : group
            }
            expanded
          >
            {Object.keys(this.state.info[group])
              .filter(a => !['label', 'typeid'].includes(a))
              .map((a, aIndex) => (
                <InfoElement
                  key={aIndex}
                  label={
                    this.state.info[group][a].label
                      ? this.state.info[group][a].label
                      : a
                  }
                  value={getValue(this.state.info[group][a])}
                  valuePath={
                    this.state.info[group][a].valuePath
                      ? this.state.info[group][a].valuePath
                      : undefined
                  }
                  alarm={infoAlarmState(this.state.info[group][a])}
                  alarmPath={
                    this.state.info[group][a].alarmStatePath
                      ? this.state.info[group][a].alarmStatePath
                      : undefined
                  }
                  tag={getTag(this.state.info[group][a])}
                  blockName={this.props.blockName}
                  attributeName={this.props.attributeName}
                  disabled={this.state.info[group][a].disabled}
                  disabledFlagPath={this.state.info[group][a].disabledPath}
                  handlers={this.state.info[group][a].functions}
                  showLabel={this.state.info[group][a].showLabel}
                  infoPath={this.state.info[group][a].infoPath}
                  infoClickHandler={this.props.infoClickHandler}
                />
              ))}
          </GroupExpander>
        ))}
      </div>
    );
  }
}

InfoDetails.propTypes = {
  blockName: PropTypes.string.isRequired,
  attributeName: PropTypes.string.isRequired,
  subElement: PropTypes.arrayOf(PropTypes.string),
  linkBlockName: PropTypes.string.isRequired,
  info: PropTypes.shape({}),
  infoClickHandler: PropTypes.func.isRequired,
  isMethodInfo: PropTypes.bool.isRequired,
  isLinkInfo: PropTypes.bool.isRequired,
  // isLinkInfo: PropTypes.bool.isRequired,
  isLayoutLocked: PropTypes.bool.isRequired,
};

InfoDetails.defaultProps = {
  subElement: undefined,
  info: undefined,
};

const deselectLinkAction = path => (dispatch, getState) => {
  const { selectedLinks } = getState().malcolm.layoutState;
  const matchingSelectedLink = selectedLinks.find(l =>
    l.endsWith(path[0] + idSeparator + path[1])
  );

  if (matchingSelectedLink) {
    dispatch(malcolmSelectLink(matchingSelectedLink, false));
  }
};

const mapDispatchToProps = dispatch => ({
  getBlock: path => {
    dispatch(malcolmGetAction(path));
  },
  changeInfoHandler: (path, subElement) => {
    dispatch(
      navigationActions.navigateToSubElement(path[0], path[1], subElement)
    );
  },
  infoClickHandler: (path, subElement) => {
    dispatch(navigationActions.navigateToInfo(path[0], path[1], subElement));
  },
  closeInfoHandler: (path, subElement) => {
    dispatch(navigationActions.closeInfo(path[0], path[1], subElement));
  },
  setFlag: (path, flag, state) => {
    dispatch(malcolmSetFlag(path, flag, state));
  },
  revertHandler: path => {
    dispatch(malcolmRevertAction(path));
  },
  eventHandler: (path, value) => {
    dispatch(malcolmSetFlag(path, 'pending', true));
    dispatch(malcolmPutAction(path, value));
  },
  rowRevertHandler: (path, value, row) => {
    dispatch(malcolmUpdateTable(path, value, row));
  },
  addRow: (path, row, modifier) => {
    dispatch(malcolmUpdateTable(path, { insertRow: true, modifier }, row));
  },
  moveRow: (path, row, modifier) => {
    dispatch(malcolmUpdateTable(path, { moveRow: true, modifier }, row));
  },
  unselectLink: path => {
    dispatch(deselectLinkAction(path));
  },
  clearParamState: (path, param) => {
    dispatch(malcolmClearMethodInput(path, param));
  },
  clearError: path => {
    dispatch(clearError(path));
  },
});

const mapStateToProps = state => {
  let blockName;
  let attributeName;
  let subElement;
  let layoutAttribute;
  const navLists = state.malcolm.navigation.navigationLists.slice(-3);
  if (navLists[2].navType === NavTypes.Info) {
    blockName =
      navLists[0].navType === NavTypes.Block ? navLists[0].blockMri : undefined;
    attributeName =
      navLists[1].navType === NavTypes.Attribute ? navLists[1].path : undefined;
    subElement =
      navLists[1].navType === NavTypes.Attribute && navLists[1].subElements
        ? navLists[1].subElements
        : undefined;
  }

  const showLinkInfo = navLists[2].path.endsWith('.link');
  let linkBlockName;
  if (showLinkInfo) {
    layoutAttribute = blockUtils.findAttribute(
      state.malcolm.blocks,
      state.malcolm.parentBlock,
      state.malcolm.mainAttribute
    );

    if (layoutAttribute && layoutAttribute.raw.value) {
      const blockNameIndex = layoutAttribute.raw.value.name.findIndex(
        n => n === navLists[2].linkInputBlock
      );

      blockName = layoutAttribute.raw.value.mri[blockNameIndex];
      attributeName = navLists[2].linkInputPort;
      linkBlockName = navLists[2].linkInputBlock;
    }
  }

  const attribute = blockUtils.findAttribute(
    state.malcolm.blocks,
    blockName,
    attributeName
  );

  return {
    attribute,
    blockName,
    attributeName,
    subElement,
    isLinkInfo: showLinkInfo,
    layoutAttribute: showLinkInfo ? layoutAttribute : undefined,
    isMethodInfo:
      attribute && attribute.calculated && attribute.calculated.isMethod,
    linkBlockName,
    isLayoutLocked: state.malcolm.layout && state.malcolm.layout.locked,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(InfoDetails);
