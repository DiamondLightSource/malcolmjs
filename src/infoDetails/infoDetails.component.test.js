import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import { InfoDetails } from './infoDetails.component';

describe('InfoDetails', () => {
  let shallow;
  const testInfo = {
    simpleValue: 1.0,
    inlineValue: {
      value: 'test',
      inline: true,
      label: 'inline info',
      alarmState: 2,
    },
    inlineWithPaths: {
      valuePath: 'something.on.attribute',
      inline: true,
      label: 'inline info',
      alarmStatePath: 'something.else.on.attribute',
      disabledPath: 'some.flag.somewhere',
    },
    inlineWithTag: {
      value: 'test in',
      inline: true,
      label: 'inline input',
      tag: 'widget:textinput',
    },
    groupValues: {
      simpleInGroup: 'zap',
      inlineInGroupWithPaths: {
        valuePath: 'something.on.attribute',
        inline: true,
        label: 'in-group info',
        alarmStatePath: 'something.else.on.attribute',
        disabledPath: 'some.flag.somewhere',
      },
      arbitraryObject: {
        a: 'b',
        c: 'd',
        e: {
          f: 'g',
        },
      },
    },
  };
  beforeEach(() => {
    shallow = createShallow();
  });
  it('renders correctly', () => {
    const wrapper = shallow(<InfoDetails info={testInfo} isLinkInfo={false} />);
    expect(wrapper).toMatchSnapshot();
  });
});
