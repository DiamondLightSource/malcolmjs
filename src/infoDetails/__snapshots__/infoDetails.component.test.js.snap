// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoDetails renders correctly 1`] = `
<div>
  <WithStyles(InfoElement)
    alarm={null}
    clickHandler={[Function]}
    key="simpleValue"
    label="simpleValue"
    tag="info:multiline"
    value={1}
  />
  <WithStyles(InfoElement)
    alarm={2}
    clickHandler={[Function]}
    key="inlineValue"
    label="inline info"
    tag="info:multiline"
    value="test"
  />
  <WithStyles(InfoElement)
    alarm={null}
    clickHandler={[Function]}
    key="inlineWithTag"
    label="inline input"
    tag="widget:textinput"
    value="test in"
  />
  <WithStyles(WidgetGroupExpander)
    expanded={true}
    groupName="groupValues"
    key="groupValues"
  >
    <WithStyles(InfoElement)
      alarm={null}
      key="simpleInGroup"
      label="simpleInGroup"
      tag="info:multiline"
      value="zap"
    />
    <WithStyles(InfoElement)
      alarm={0}
      key="arbitraryObject"
      label="arbitraryObject"
      tag="info:multiline"
      value="b"
    />
  </WithStyles(WidgetGroupExpander)>
</div>
`;
